### PR TITLE
OCPBUGS-33018: pkg/daemon: Ignore watch failure unless kubelet needs a rebootstrap

### PR DESCRIPTION
**Pull request by wking**
Some adjustements to the logic from 4d447c5c8f (#4106).

Factor out the common rebootstrap logic into a new function, just to avoid repeating ourselves in two places.  This part is a pure refactor, with no user-visible changes.

Significantly for users, 4d447c5c8f's `utilruntime.ErrorHandlers` handling introduced a race like:

1. Daemon is rolling an update out onto disk.
2. Kube API or DNS hiccup causes the daemon's watches, or other Kube API requests, to fail.
3. `utilruntime.ErrorHandlers` feeds that failure into `errCh`.
4. `stopCh` set to trigger graceful shutdowns in the child goroutines.
5. Child goroutine performing the update moves into its deferred rollbacks (for example, see code around [`error rolling back files writes`][rollback]).
6. But before that rollback completes, the main goroutine returns `ErrAuxiliary`, and the container exits 255 without finishing the rollback.
7. Kubelet launches a replacement container.
8. Replacement container is upset at the interrupted cleanup, complains `content mismatch for file...` about some of the incoming-but-not-rolled-back content, and exits.
9. Return to step 7, looping forever.

But we ignored these `utilruntime.ErrorHandlers` errors before 4d447c5c8f, without trouble beyond responding to `api-int` Certificate Authority rotations.  We still want to rebootstrap when we have `api-int` trouble, but I'm not gating both the rebootstrap and the `ErrAuxiliary` return on `deferKubeletRestart`.  So now:

1. Daemon is rolling an update out onto disk.
2. Kube API or DNS hiccup causes the daemon's watches, or other Kube API requests, to fail.
3. `utilruntime.ErrorHandlers` feeds that failure into `errCh`.
4. Daemon knows it does not have a pending kubelet rebootstrap, and it ignores the error.

Or we can have:

...
4. Daemon knows it needs to rebootstrap the kubelet, so it does.
5. Daemon returns `ErrAuxiliary`, and the container exits 255.

There's still room for a partial-rollback race there if we have an `api-int` certificate rotation in the works while the Kube API hiccups while we're rolling update files out to disk, because the current goroutine handling launches children like:

```go
go wait.Until(dn.worker, time.Second, stopCh)
```

so there's no mechanism for the canceled child goroutine to tell the parent "got your message, and I've finished my graceful shutdown". But this commit means we are only exposed when there's an ongoing `api-int` rotation, and that reduces our risk significantly (they're less common than generic "any kind of Kube API hiccup matching the daemon's substrings").  [MCO-1154](https://issues.redhat.com/browse/MCO-1154) is up to track providing that child-to-parent return path.

[rollback]: https://github.com/openshift/machine-config-operator/blob/9b99954aa3c1d5592cf13c02a7b6e10ac142d9a3/pkg/daemon/update.go#L1168-L1176

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -1386,36 +1368,41 @@ func (dn *Daemon) Run(stopCh <-chan struct{}, exitCh <-chan error, errCh chan er
...
 				}
 				return ErrAuxiliary
 			}
 		}
 	}
 }
 
+func (dn *Daemon) kubeletRebootstrap(ctx context.Context) error {
+	dn.deferKubeletRestart = false
+	err = os.Remove("/var/lib/kubelet/kubeconfig")
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:1384]: I think the build is failing on this not being a `:=`

Comment (wking) [reply pkg/daemon/daemon.go]: Thanks!  Fixed in 1a8ab7752 -> 66cf5a193.

**Comment by wking** [comment]
Running one to see if this works:

/payload-job periodic-ci-openshift-release-master-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6

If that shows us using the target payload we expect, we probably want to run a [`/payload-aggregate ...`][1] because [the periodic is only mostly dead, not completely dead][2], and we don't want to misread a got-lucky-and-passed racy win as "actually fixed this issue".

[1]: https://docs.ci.openshift.org/docs/release-oversight/payload-testing/
[2]: https://testgrid.k8s.io/redhat-openshift-ocp-release-4.16-informing#periodic-ci-openshift-release-master-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -1383,39 +1365,45 @@ func (dn *Daemon) Run(stopCh <-chan struct{}, exitCh <-chan error, errCh chan er
...
-							return false, nil
-						} else if err != nil {
-							return false, fmt.Errorf("unexpected error reading kubeconfig file, %v", err)
-						}
-
-						return true, nil
-					}); err != nil {
-						return fmt.Errorf("something went wrong while waiting for kubeconfig file to generate: %v", err)
-					}
+			if dn.deferKubeletRestart && (strings.Contains(strings.ToLower(err.Error()), "failed to watch") || strings.Contains(strings.ToLower(err.Error()), "unknown authority") || strings.Contains(strings.ToLower(err.Error()), "error on the server")) {
```
Comment (yuqi-zhang) [inline pkg/daemon/daemon.go:1371]: For a second I thought this was doing the same thing, but the key difference is that `ErrAuxiliary` gets wrapped into a kubelet restart. I think this should be safe so let's give it a try

**Review by yuqi-zhang — APPROVED** [review]

**Comment by wking** [comment]
[failed][1], [with][2]:

```
INFO[2024-04-25T23:12:42Z] Resolved source https://github.com/openshift/machine-config-operator to master@9b99954a, merging: #4337 66cf5a19 @wking
```

and:

```
event happened 663 times, something is wrong: namespace/openshift-config-operator node/master-0.ostest.test.metalkube.org pod/openshift-config-operator-55d4f665c8-l9l66 hmsg/b037c7d6d4 - reason/BackOff Back-off pulling image "registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef" (04:11:53Z) result=reject }
```

Not sure what's going on with the config operator.  [And gather-extra failed to collect pod logs][3].  And no bytes in the pod logs in the must-gather:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664/artifacts/e2e-metal-ipi-upgrade-ovn-ipv6/gather-must-gather/artifacts/must-gather.tar | tar -tvz | grep '/openshift-config-operator.*/logs/'
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-api/openshift-api/logs/current.insecure.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-api/openshift-api/logs/current.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-api/openshift-api/logs/previous.insecure.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-api/openshift-api/logs/previous.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-config-operator/openshift-config-operator/logs/current.insecure.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-config-operator/openshift-config-operator/logs/current.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-config-operator/openshift-config-operator/logs/previous.insecure.log
-rw------- 1012900000/root        0 2024-04-25 21:20 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-0a2db4d71d7957fc2a92bc07c98918f169650dc0a6d040f40a26313e98bba9c3/namespaces/openshift-config-operator/pods/openshift-config-operator-55d4f665c8-l9l66/openshift-config-operator/openshift-config-operator/logs/previous.log
```

Ah, because we failed to pull that image:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664/artifacts/e2e-metal-ipi-upgrade-ovn-ipv6/gather-extra/artifacts/pods.json | jq -r '.items[] | select(.metadata.name == "openshift-config-operator-55d4f665c8-l9l66").status.initContainerStatuses[]'
{
  "image": "registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef",
  "imageID": "",
  "lastState": {},
  "name": "openshift-api",
  "ready": false,
  "restartCount": 0,
  "started": false,
  "state": {
    "waiting": {
      "message": "Back-off pulling image \"registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef\"",
      "reason": "ImagePullBackOff"
    }
  }
}
```

Because networking fell over?  Or some kind of proxy thing?  I dunno:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664/artifacts/e2e-metal-ipi-upgrade-ovn-ipv6/gather-extra/artifacts/events.json | jq -r '.items[] | select(.reason == "Failed" and (.message | contains("3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef"))).message'
Failed to pull image "registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef": (Mirrors also failed: [virthost.ostest.test.metalkube.org:5000/localimages/local-upgrade-image@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef: reading manifest sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef in virthost.ostest.test.metalkube.org:5000/localimages/local-upgrade-image: manifest unknown]): registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef: pinging container registry registry.build05.ci.openshift.org: Get "https://registry.build05.ci.openshift.org/v2/": dial tcp 54.145.168.129:443: connect: network is unreachable
Failed to pull image "registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef": (Mirrors also failed: [virthost.ostest.test.metalkube.org:5000/localimages/local-upgrade-image@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef: reading manifest sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef in virthost.ostest.test.metalkube.org:5000/localimages/local-upgrade-image: manifest unknown]): registry.build05.ci.openshift.org/ci-op-s98n4wp9/stable@sha256:3ac58db589768e083d2a4378dec14a7ebc2329e8629f7bc78ab1a5c691a6e0ef: pinging container registry registry.build05.ci.openshift.org: Get "https://registry.build05.ci.openshift.org/v2/": dial tcp 54.165.220.45:443: connect: network is unreachable
```

But:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664/artifacts/e2e-metal-ipi-upgrade-ovn-ipv6/gather-extra/artifacts/clusterversion.json | jq -r '.items[].status.history[] | .startedTime + " " + (.completionTime // "-") + " " + .state + " " + .version'
2024-04-26T01:41:22Z - Partial 4.16.0-0.ci.test-2024-04-25-231905-ci-op-s98n4wp9-latest
2024-04-26T00:10:33Z 2024-04-26T01:05:53Z Completed 4.15.10
```

so looks like it is doing the 4.15 -> freshly-built 4.16 that we want.  I'll run another:

/payload-job periodic-ci-openshift-release-master-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664
[2]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664#1:build-log.txt%3A3
[3]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/openshift-machine-config-operator-4337-nightly-4.16-upgrade-from-stable-4.15-e2e-metal-ipi-upgrade-ovn-ipv6/1783635215088881664/artifacts/e2e-metal-ipi-upgrade-ovn-ipv6/gather-extra/artifacts/

**Comment by yuqi-zhang** [comment]
We can't seem to pre-merge test this via the payload and it's a safe enough change.

/lgtm
/override ci/prow/e2e-hypershift

**Pull request merged**